### PR TITLE
refactor(test_ode.py): regression tests

### PR DESCRIPTION
**Pull request by Mohitbalwani26**
#### References to other Issues or PRs

SEE #18377

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/solvers/ode/tests/test_single.py [hunk 1]
```diff
@@ -598,9 +571,37 @@ def _get_examples_ode_sol_factorable():
         'sol': [], #currently dsolve doesn't return any solution for this example
         'XFAIL': ['factorable']
     },
+
+    #Below examples were added for the issue: https://github.com/sympy/sympy/issues/15889
+    'fact_12': {
```
Comment (oscarbenjamin) [inline sympy/solvers/ode/tests/test_single.py:576]: It's better to say which examples explicitly e.g. `fact_12 through fact_16`.

**Comment by oscarbenjamin** [comment]
This looks fine. Let me know when you think it's ready to be merged.

File: sympy/solvers/ode/tests/test_single.py (hunk 1 above)
Comment (Mohitbalwani26) [reply sympy/solvers/ode/tests/test_single.py:576]: okay i will do it next commit.

**Comment by Mohitbalwani26** [comment]
> This looks fine. Let me know when you think it's ready to be merged.

After refactoring a few more functions i will let you know for sure.

File: sympy/solvers/ode/tests/test_single.py [hunk 2]
```diff
@@ -293,33 +293,6 @@ def test_factorable():
     _ode_solver_test(_get_examples_ode_sol_factorable())
 
 
-def test_issue_15889():
-    eq = exp(f(x).diff(x))-f(x)**2
-    sol = Eq(NonElementaryIntegral(1/log(y**2), (y, f(x))), C1 + x)
-    assert sol.dummy_eq(dsolve(eq))
-    assert checkodesol(eq, sol) == (True, 0)
```
Comment (Mohitbalwani26) [inline sympy/solvers/ode/tests/test_single.py:300]: @oscarbenjamin can you help how should this example be handled?

Comment (oscarbenjamin) [reply sympy/solvers/ode/tests/test_single.py:300]: I guess that if the expected solution (i.e. `sol` above) has any `Dummy` symbols we can use `dummy_eq` to compare the solutions:
```
In [63]: eq = exp(f(x).diff(x))-f(x)**2                                                                                           

In [64]: sol = dsolve(eq, f(x))                                                                                                   

In [65]: sol.has(Dummy)                                                                                                           
Out[65]: True
```
We should check for Dummy in the expected solution though, *not* the result from `dsolve`.

Comment (Mohitbalwani26) [reply sympy/solvers/ode/tests/test_single.py:300]: when we look at expected solution `eq = Eq(NonElementaryIntegral(1/log(y**2), (y, f(x))), C1 + x)` so here `eq.has(Dummy)` will be false.

should i define a dummy variable and write the equation?

Comment (oscarbenjamin) [reply sympy/solvers/ode/tests/test_single.py:300]: > should i define a dummy variable and write the equation?

Yes. The convention can be that using a `Dummy` in the "expected" solution means it should be compared using `dummy_eq` in the tests.

**Comment by Mohitbalwani26** [comment]
@oscarbenjamin i didn't get the reason why travis is failing?

**Comment by oscarbenjamin** [comment]
It's just a network failure. The tests will pass next time they run.

Normally I would be able to restart a test like that but the restart button seems to have disappeared on Travis...

**Comment by jksuom** [comment]
I've restarted it. Maybe you were logged out of Travis.

**Comment by Mohitbalwani26** [comment]
@oscarbenjamin can you review this now?

File: sympy/solvers/ode/tests/test_single.py [hunk 3]
```diff
@@ -163,6 +163,13 @@ def _test_all_hints(runxfail=False):
         _test_all_examples_for_one_hint(our_hint, all_examples, runxfail)
 
 
+def _test_dummy_sol(expected_sol,dsolve_sol):
+    if type(dsolve_sol)==list:
+        return any(expected_sol.dummy_eq(sub_dsol) for sub_dsol in dsolve_sol)
```
Comment (oscarbenjamin) [inline sympy/solvers/ode/tests/test_single.py:168]: We need to confirm that all solutions from expected_sol match something from dsolve_sol and vice versa.

**Comment by oscarbenjamin** [comment]
I think that `_test_particular_example` is getting very messy. It's hard for me to tell by looking at it that it is actually correct. I think it can be simplified. Firstly I'm not sure why the main if/else is there at all.

File: sympy/solvers/ode/tests/test_single.py (hunk 3 above)
Comment (Mohitbalwani26) [reply sympy/solvers/ode/tests/test_single.py:168]: yes it is tested.

**Comment by Mohitbalwani26** [comment]
> I think that `_test_particular_example` is getting very messy. It's hard for me to tell by looking at it that it is actually correct. I think it can be simplified. 

which part according to you is messy? 

>Firstly I'm not sure why the main if/else is there at all.
when we run the tests there is a message for debugging
```
from sympy.solvers.ode.tests.test_single import _test_particular_example

_test_particular_example('{hint}', example_name = '{example}')
```
so this function can be run by the user as well.
basically this function is used by both users and internal functions that's why parameters are handled.

**Comment by oscarbenjamin** [comment]
> basically this function is used by both users and internal functions that's why parameters are handled.

A better way of handling this is to have two functions. They shouldn't duplicate the implementation but e.g. one can prepare the arguments for the other.

**Comment by Mohitbalwani26** [comment]
> > basically this function is used by both users and internal functions that's why parameters are handled.
> 
> A better way of handling this is to have two functions. They shouldn't duplicate the implementation but e.g. one can prepare the arguments for the other.

can you elaborate your idea because i don't think there will better option than current implementation of handling them?

**Comment by oscarbenjamin** [comment]
> can you elaborate your idea because i don't think there will better option than current implementation of handling them?

There's always a better way :)

The two branches of the main `if/else` are basically two separate functions mashed together. That should be two different functions.

**Comment by oscarbenjamin** [comment]
I think it would make more sense if the `_test_an_example` function calls `test_particular_example` or if it outputs the code to create and solve the example directly.

**Comment by Mohitbalwani26** [comment]
> I think it would make more sense if the `_test_an_example` function calls `test_particular_example` or if it outputs the code to create and solve the example directly.

```python3

def _test_an_example(our_hint, example_name):
    all_examples = _get_all_examples()
    for example in all_examples:
        if example['example_name'] == example_name:
            _test_particular_example(our_hint, example)

```
like this?

**Comment by oscarbenjamin** [comment]
Yes, that looks better

**Comment by Mohitbalwani26** [comment]
I have updated the changes. Does this pr look good to merge now?

**Comment by Mohitbalwani26** [comment]
@oscarbenjamin do you want to suggest any changes for this?

**Comment by oscarbenjamin** [comment]
Looks good.

**Comment by oscarbenjamin** [comment]
Thanks!

**Comment by Mohitbalwani26** [comment]
@oscarbenjamin how should the tests marked as `@slow` be handled?

**Comment by oscarbenjamin** [comment]
I think individual examples should be marked as slow if they are slow and then there should a `test_` functions marked as slow that only run the slow tests. Probably slow should be an option to some of the test examples functions.

**Comment by Mohitbalwani26** [comment]
so on what basis should I mark particular examples as slow because in master branch whole functions are marked like that?

**Comment by oscarbenjamin** [comment]
We should time them all and give a sorted list of their times. There could be a threshold in there somewhere. Actually it would be even better if we can distinguish between whether dsolve or checkodesol is slow.

**Pull request merged**